Add TLSSocket greentea tests.

michalpasztamobica opened the pull request:
### Description

TLSSocket tests based on existing TCPSocket greentea tests and TLSSocket icetea tests.
The tests are only passing with `"rtos.main-thread-stack-size": 8192` (6144 if we take out the SIMULTANEOUS test).
HTTPS_GET test is commented as it fails, perhaps due to connectivity reasons - I don't want to break the CI.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [x] Test update
    [ ] Breaking change

### Reviewers

@SeppoTakalo 
@KariHaapalehto 
@VeijoPesonen 
@mtomczykmobica

michalpasztamobica committed:
Add TLSSocket greentea tests.

TLSSocket tests based on existing TCPSocket greentea tests and TLSSocket icetea tests.

ciarmcom commented:
@michalpasztamobica, thank you for your changes.
@VeijoPesonen @SeppoTakalo @KariHaapalehto @mtomczykmobica @ARMmbed/mbed-os-ipcore @ARMmbed/mbed-os-test @ARMmbed/mbed-os-maintainers please review.

VeijoPesonen commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+    sock.set_timeout(0);
+    while (true) {
+        err = sock.recv(NULL, 0);
+        if (err == NSAPI_ERROR_WOULD_BLOCK) {
```
Adding the header would be good

`#include "features/netsocket/nsapi_types.h"`

VeijoPesonen replied:
Never mind, `#include "mbed.h"` is sufficient in test cases.

mtomczykmobica reviewed (approved):
Looks good.

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+                               "X4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlG\n" \
+                               "PfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6\n" \
+                               "KOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==\n" \
+                               "-----END CERTIFICATE-----\n";
```
The `\` at the end of the line is not required as C/C++ automatically appends strings together.

However, this is minor.. don't mind if it stays..

This is left from previously the cert being `#define CERTIFICATE` where it is a multiline macro.

michalpasztamobica replied:
Fixed, thanks for pointing this out.

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_echotest_burst.cpp:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        if (bt_left != 0) {
+            drop_bad_packets(*sock, 0);
```
This can only happen if `recv()` returns error and if it does, we should not try to read again.

I prefer this `drop_bad_packets()` to be removed. Just mark the test failed (with error message, if possible).

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+        }
+    }
+    sock.set_timeout(orig_timeout);
+}
```
I don't understand the usage of this function..

If the socket fails, we can just close it. There is no need to drain the pipes for all unread data. Closing the socket should properly drop all unread data.

michalpasztamobica replied:
You are right. I can see this function being useful only in case we want to flush the packets without closing the socket. I will leave the function in, but I will remove its usage (as you suggested above).

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+{
+    SocketAddress tls_addr;
+
+    get_interface()->gethostbyname(MBED_CONF_APP_ECHO_SERVER_ADDR, &tls_addr);
```
Probably a minor thing, but now that we use `NetworkInterface::get_default_instance()` can we refactor all these `get_interface()` call to  `NetworkInterface::get_default_instance()->`

I feel like `get_interface()` is also a residue from time when default network interface did not exist

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+    if (err != NSAPI_ERROR_OK) {
+        printf("Error from sock.connect: %d\n", err);
+        return err;
+    }
```
This function is copy&paste from previous, only change is the port address.
Can the port address be a parameter to this and just use the same function?

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,231 @@
+    Case("TLSSOCKET_OPEN_TWICE", TLSSOCKET_OPEN_TWICE),
+    Case("TLSSOCKET_OPEN_LIMIT", TLSSOCKET_OPEN_LIMIT),
+    Case("TLSSOCKET_OPEN_DESTRUCT", TLSSOCKET_OPEN_DESTRUCT),
+//    Case("TLSSOCKET_HTTPS_GET", TLSSOCKET_HTTPS_GET), // Fails!
```
`// Fails!` does not sound very good on public example.

Maybe "work-in-progress, unstable" or something like that.

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_https_get.cpp:
```diff
@@ -0,0 +1,35 @@
+    TLSSocket sock;
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.open(get_interface()));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.set_root_ca_cert(tls_global::cert));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.connect("os.mbed.com", 443));
```
Hmm...
Our root certificate is the Letsencrypt.org one..
The `os.mbed.com` uses different root.. Therefore this test cannot pass.

We could for example use some other server, with same certificate, or just drop the test case.

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_bind_port_fail.cpp:
```diff
@@ -0,0 +1,67 @@
+        TEST_FAIL();
+    }
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock2->open(get_interface()));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_PARAMETER, sock2->bind(sockAddr));
```
I'm wondering whether these `Socket::bind()` testcases makes any sense on TLSSocket as the code is basically just `transport->bind(...)` and the transport socket is already tested.

Maybe we can drop the bind() test cases.

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_echotest.cpp:
```diff
@@ -0,0 +1,195 @@
+                TEST_FAIL();
+                break;
+            }
+            wait(1);
```
Can we instead wait for the sigio signal to happen?

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_echotest.cpp:
```diff
@@ -0,0 +1,195 @@
+
+    tlssocket_connect_to_echo_srv(*sock);
+    sock->set_blocking(false);
+    sock->sigio(callback(_sigio_handler, ThisThread::get_id()));
```
It would be nice to deliver this signal to backround receiver as well.

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_echotest.cpp:
```diff
@@ -0,0 +1,195 @@
+                            tls_global::TLS_OS_STACK_SIZE,
+                            stack_mem,
+                            "receiver");
+        TEST_ASSERT_EQUAL(osOK, thread->start(callback(tlssocket_echotest_nonblock_receiver, &pkt_s)));
```
New thread per packet?

Sounds a bit overkill...


We could push event to global event loop every time we receive sigio(), and that event could the read until WOULD_BLOCK is received.. but if all data is received, it could set some global event flag to indicate that all data is read.

michalpasztamobica replied:
Just to make sure we have a common understanding of what is going on in this test...

Indeed, a thread is created for every new packet size. But there is always just one receiver thread running at a time (tx_sem.wait will block the main thread's for loop until receiver thread exits). The receiver thread already contains a loop, which is controller with SIGIO (as per the previous comment :) ) and returns once all data is received.
So the only change would be to trigger the receiver's loop using a global event loop instead of triggering it using thread creation - is that right? Or is there some further optimisation which I failed to notice?

SeppoTakalo commented on TESTS/netsocket/README.md:
```diff
@@ -102,35 +102,41 @@ time        stream  tcp6    nowait  root    internal
+Nmap done: 1 IP address (1 host up) scanned in 0.10 seconds
 
-$ sudo nmap -sT -p7,9,13,37 echo.mbedcloudtesting.com
+$ sudo nmap -sT -p7,9,13,37,2007,2009,2013 echo.mbedcloudtesting.com
```
I can see that there was a error on previous documentation.
This was supposed to do UDP scan, and therefore required `sudo`

Can you change it to do
`sudo nmap -sU` instead?

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_endpoint_close.cpp:
```diff
@@ -0,0 +1,91 @@
+ * limitations under the License.
+ */
+
+#include "tls_tests.h""
```
Douple "" at the end of line

SeppoTakalo commented on TESTS/netsocket/tls/tlssocket_endpoint_close.cpp:
```diff
@@ -0,0 +1,91 @@
+        TEST_FAIL();
+        return;
+    }
+    sock.sigio(callback(_sigio_handler, ThisThread::get_id()));
```
Not sure, but can we actually construct `Callback` object with two parameters?

Like:
`sock.sigio(callback(osSignalSet, ThisThread::get_id(), SIGNAL_SIGIO);`

So we would not need the `_sigio_handler()` wrapper.

michalpasztamobica replied:
There is no such template :(

SeppoTakalo reviewed (changes requested):
Just a minor changes proposed.

Otherwise looks good.

VeijoPesonen commented on TESTS/netsocket/tls/tlssocket_bind_unopened.cpp:
```diff
@@ -0,0 +1,56 @@
+
+    TLSSocket *sock = new TLSSocket;
+    if (!sock) {
+        TEST_FAIL();
```
TEST_FAIL won't abort execution so adding return-statement would be good here.

michalpasztamobica replied:
Good point, but I removed the bind tests entirely. 
However, these same comments apply also to tcp tests.

VeijoPesonen commented on TESTS/netsocket/tls/tlssocket_no_cert.cpp:
```diff
@@ -0,0 +1,33 @@
+{
+    TLSSocket sock;
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.open(get_interface()));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_AUTH_FAILURE, sock.connect(MBED_CONF_APP_ECHO_SERVER_ADDR, 2007));
```
Would be good to take the port number from app config json-file as this is non-standard port

VeijoPesonen commented on TESTS/netsocket/tls/tlssocket_send_unconnected.cpp:
```diff
@@ -0,0 +1,34 @@
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.open(get_interface()));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.set_root_ca_cert(tls_global::cert));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_NO_CONNECTION, sock.send("12345", 5));
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.close());
```
Should NSAPI_ERROR_NO_CONNECTION be also here the only accepted return code?

SeppoTakalo replied:
Hmm.. Actually, you can call `Socket::close()` before connecting it.

Close is originally opposite of `Socket::open()` but in fact it serves as disconnect call as well for TCPSocket.

Confusion partially comes from POSIX where `socket()` creates a socket and `close()` does both, disconnection and destruction of the file handle.

SeppoTakalo commented:
@ARMmbed/mbed-os-maintainers @ARMmbed/mbed-os-core Hi guys, do you know that is there any way we could force main stack size to be raised to 8 kB just for these TLSSocket tests?

Currently that is roughly the amount of stack that TLS handshake requires. Therefore we need to resolve the stack issue before merging?

For example, can I manually create a thread that has 8kB of stack, and launch Greentea from there?

SeppoTakalo commented:
@michalpasztamobica Can you try for example following:
```

static int run_test(voi) {
    return !Harness::run(specification);
}

static char stack_mem[8192];
int main()
{
    Thread *th = new Thread(osPriorityNormal,
                        8192,
                         stack_mem,
                         "tls_gt_thread");
    th->start(run_test);
    return th->join();
}
```

If that works, we could perhaps use that to force bigger stack for TLS thread.

michalpasztamobica commented:
@SeppoTakalo , your idea worked. Thanks!
I also added the echo-server-port-tls and echo-server-discard-port-tls definitions in tools/test_configs/*.json files.
If this is approved we can merge safely.

SeppoTakalo reviewed (approved):
Now looks OK

OPpuolitaival reviewed (approved):
Looks good

michalpasztamobica committed:
Publish TLSSocket test documentation

The whole README.md had to be updated to match the internal Confluence documentation, which can now be locked. In the process I also updated any spotted mistakes in tests documentation, removed the obsolete TCPServer tests documentation and added a test which was missing from UDPSocket's main.

michalpasztamobica commented:
@SeppoTakalo , @VeijoPesonen , I updated the README.md file with the TLSSocket test documentation and also amended some mistakes I spotted. I did not open a new PR, as the changes were actually relevant to this PR and it still wasn't merged. 
Please, review just that one file once more.

michalpasztamobica commented:
I just pushed a minor update of the README.md file, to get the test documentation right after the update in #9392 . The test code itself is already correct.
@cmonr , @0xc0170 , I think this PR has already been reviewed by all persons interested :)

michalpasztamobica commented:
Pushed a fix for Travis failures in astyle.

0xc0170 commented:
Note, Travis might fail, there is a fix landing master soon (another dependency issue).

michalpasztamobica commented:
Thanks, @0xc0170 , but these were actual astyle failures. I don't know how come they haven't popped up earlier...

michalpasztamobica commented:
Rebased on top of the travis fix.

michalpasztamobica commented:
It seems to me the travis failure is some kind of environment issue...
```
$ astyle --version
The program 'astyle' is currently not installed. To run 'astyle' please ask your administrator to install the package 'astyle'
The command "astyle --version" failed and exited with 127 during .
```

0xc0170 commented:
AStyle job restarted

0xc0170 commented:
Travis fixed, will go to CI

0xc0170 commented:
CI started

0xc0170 commented:
> TESTS/netsocket/tls/main.cpp",199  Error[Pe289]: no instance of constructor "utest::v1::Specification::Specification" matches the argument list

Failures in the build

michalpasztamobica commented:
I think this is the issue:
```
13:02:09 [UBLOX_C027:GCC_ARM]         [Error] tls_tests.h@24,23: variable or field 'drop_bad_packets' declared void
13:02:09 [UBLOX_C027:GCC_ARM]         [Error] tls_tests.h@24,23: 'TLSSocket' was not declared in this scope
13:02:09 [UBLOX_C027:GCC_ARM]         [Error] tls_tests.h@24,34: 'sock' was not declared in this scope
13:02:09 [UBLOX_C027:GCC_ARM]         [Error] tls_tests.h@24,40: expected primary-expression before 'int'
```
I got over it locally by adding #include mbed.h in the beginning of tls_tests.h, but this trick does not work for CI. 
@SeppoTakalo , @0xc0170 , is there some place I need to mention a new greentea test directory for the build system to work fine? I was trying to find out  if tcp or udp were added somewhere, but couldn't find anything...
EDIT:
Or in other words... how come TESTS/netsocket/tcp/tcp_tests.h can use NetworkInterface type, without any includes? TESTS/netsocket/tls/tls_tests.h cannot do that without including mbed.h. The main.cpp files for both directories have identical include lists.

0xc0170 commented:
I would start not including mbed.h in that header file first but include what it needs (there are only few declarations there should be staight forward to replace?).

> Or in other words... how come TESTS/netsocket/tcp/tcp_tests.h can use NetworkInterface type, without any includes? 

You can forward declare it

michalpasztamobica commented:
Ok, I think I found the issue. The other tests always add `#include tcp_tests.h` after all other includes. I did the same for tls and it compiled then. No idea how it wasn't complaining before...

michalpasztamobica commented:
I did not modify the offending files:
```
platform/mbed_rtc_time.h
=================================
Errors: 
101:/** retarget
113:/** retarget
103: secondsa
115: secondsa
...
Total Errors Found: 4
If any of the failed words should be considered valid please add them to the ignore.en.pws file found in tools/test/scripts/doxy-spellchecker between the _code_ and _doxy_ tags.
The command "./tools/test/travis-ci/doxy-spellchecker/spell.sh platform" exited with 1.
```

@0xc0170 , should I fix this?

0xc0170 commented:
Fix is waiting for test to finish, should be within a hour on master

cmonr commented:
**NOTE:** This PR has now been rebased.

If this was made in error, feel free to force-push your local branch to restore the PR.

cmonr commented:
CI started

cmonr commented:
Just merged a couple of PRs. This is gonna need a rebase resolution.

michalpasztamobica commented:
Rebased and conflicts resolved.

michalpasztamobica commented:
@0xc0170 , @cmonr , could you please start the CI here, so we make sure the compilation runs fine? I had to reenter the most recent changes after last night's rebase. Code compiles fine for me locally, but I already saw the CI act differently. Perhaps the compilation order makes a difference here - with the previous code, if the tls_tests.h got included after all the other includes, then no errors were thrown.

0xc0170 commented:
Ci needs an update, will restart this job right after

0xc0170 commented:
CI restarted

michalpasztamobica commented:
@cmonr , @0xc0170 , sorry that the build keeps failing. I tried both on Linux and Windows and it never fails for me locally. I am guessing that the issue is still the build order. It seems locally main.cpp gets compiled at the very end when all possible includes are already somehow parsed and there are no complains from the compiler about the fact that NetworkInterface is used without any headers that would contain declaration. In the logs I can see that the main.cpp compiles at 6,2% (rather at the beggining) and the failure seems to be caused by misinterpretation of the line containing the NetworkInterface usage.
 I added the following include:
```
TESTS/netsocket/tls/tls_tests.h
@@ -18,6 +18,8 @@
 #ifndef TLS_TESTS_H
 #define TLS_TESTS_H
 
+#include "TLSSocket.h"
+
 NetworkInterface *get_interface();
```
Would you please trigger the build again so we can see if that fixes the issue?

0xc0170 commented:
CI restarted

michalpasztamobica commented:
Dear @OPpuolitaival,
Once this PR builds in CI and gets merged it might introduce an issue similar to the one we've seen in a [PR that already went in and broke CI](https://github.com/ARMmbed/mbed-os/pull/9376), which was [fixed by adding the relevant parameters](https://github.com/ARMmbed/mbed-os-ci/pull/408) in the CI's config files.
I believe this PR might have similar effect - I've added to parameters: 
```
        "echo-server-port-tls" : {
            "help" : "Port of echo server for TLS",
            "value" : "2007"
        },
        "echo-server-discard-port-tls" : {
            "help" : "Discard port of echo server for TLS",
            "value" : "2009"
        }
```

OPpuolitaival commented:
@michalpasztamobica This seems to be a bit different. In the case where you point the manually created configuration did not have all the fields in place. But default configuration should be able to build. Can you try to execute `mbed test --compile ` for those targets which are listed end of this comment and see if you can compile those with default configuration.

Error message:
[ARCH_PRO:ARM]         [Error] tls_tests.h@24,0:  #70: incomplete type is not allowed
[ARCH_PRO:ARM]         [Error] tls_tests.h@24,0:  #20: identifier "TLSSocket" is undefined
[ARCH_PRO:ARM]         [Error] tls_tests.h@24,0:  #20: identifier "sock" is undefined
[ARCH_PRO:ARM]         [Error] tls_tests.h@24,0:  #18: expected a ")"
[ARCH_PRO:ARM]         [Error] main.cpp@76,0:  #247: variable "drop_bad_packets"  has already been defined
[ARCH_PRO:ARM]         [Error] main.cpp@76,0:  #70: incomplete type is not allowed
[ARCH_PRO:ARM]         [Error] main.cpp@76,0:  #18: expected a ")"
[ARCH_PRO:ARM]         [Error] main.cpp@77,0:  #65: expected a ";"
[ARCH_PRO:ARM]         [Warning] main.cpp@130,0:  #12-D: parsing restarts here after previous syntax error
[ARCH_PRO:ARM]         [Error] main.cpp@131,0:  #169: expected a declaration
[ARCH_PRO:ARM]         [Warning] main.cpp@197,0:  #12-D: parsing restarts here after previous syntax error
[ARCH_PRO:ARM]         [Error] main.cpp@199,0:  #20: identifier "greentea_setup" is undefined
[ARCH_PRO:ARM]         [Error] main.cpp@199,0:  #20: identifier "cases" is undefined
[ARCH_PRO:ARM]         [Error] main.cpp@199,0:  #20: identifier "greentea_teardown" is undefined
[ARCH_PRO:ARM]         [Error] main.cpp@199,0:  #289: no instance of constructor "utest::v1::Specification::Specification"  matches the argument list
[ARCH_PRO:ARM]         [DEBUG] Return: 1
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/tls_tests.h", line 24: Error:  #70: incomplete type is not allowed
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/tls_tests.h", line 24: Error:  #20: identifier "TLSSocket" is undefined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/tls_tests.h", line 24: Error:  #20: identifier "sock" is undefined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/tls_tests.h", line 24: Error:  #18: expected a ")"
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 76: Error:  #247: variable "drop_bad_packets"  has already been defined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 76: Error:  #70: incomplete type is not allowed
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 76: Error:  #18: expected a ")"
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 77: Error:  #65: expected a ";"
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 130: Warning:  #12-D: parsing restarts here after previous syntax error
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 131: Error:  #169: expected a declaration
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 197: Warning:  #12-D: parsing restarts here after previous syntax error
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 199: Error:  #20: identifier "greentea_setup" is undefined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 199: Error:  #20: identifier "cases" is undefined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 199: Error:  #20: identifier "greentea_teardown" is undefined
[ARCH_PRO:ARM]         [DEBUG] Output: "./TESTS/netsocket/tls/main.cpp", line 199: Error:  #289: no instance of constructor "utest::v1::Specification::Specification"  matches the argument list
[ARCH_PRO:ARM]         [DEBUG] Output:             argument types are: (<error-type>, <error-type>, <error-type>, const utest::v1::handlers_t)
[ARCH_PRO:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors


And quite many platform is failing for same error:

$grep "13 erro" consoleText.txt
[ARCH_PRO:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M0P:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M0:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M7:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M4:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M3:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[GD32_F307VG:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC4088_DM:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC1769:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC4088:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC1768:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[MIMXRT1050_EVK:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[NUCLEO_F207ZG:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[RZ_A1H:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 3 warnings, 13 errors
[UBLOX_C027:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[VK_RZ_A1H:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 3 warnings, 13 errors
[FVP_MPS2_M7:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M0:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M0P:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M4:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[ARCH_PRO:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[FVP_MPS2_M3:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC1769:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC1768:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC4088:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[LPC4088_DM:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[GD32_F307VG:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[VK_RZ_A1H:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 3 warnings, 13 errors
[NUCLEO_F207ZG:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[UBLOX_C027:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors
[RZ_A1H:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 3 warnings, 13 errors
[MIMXRT1050_EVK:ARM]         [DEBUG] Output: ./TESTS/netsocket/tls/main.cpp: 2 warnings, 13 errors

michalpasztamobica commented:
@OPpuolitaival , you're right, I got the error locally when I compiled for ARCH_PRO. I will try to debug this issue. Thanks for the hint!

michalpasztamobica commented:
The problem was that TLSSocket definitions are all wrapped in #if defined(MBEDTLS_SSL_CLI_C) in order not tocompile them on platform which do not support mbedtls.

I wrapped the tls tests with the same macro.
I checked three failing architectures with three different compilers and they now build fine.
I also made sure that the other architectures (like K64F) will build the tests and pass them.

Once again, thank you for pointing out where the issue lies, @OPpuolitaival

michalpasztamobica commented:
Fixed the missing whitespace which Travis pointed out.

SeppoTakalo commented on TESTS/netsocket/tls/main.cpp:
```diff
@@ -0,0 +1,223 @@
+#include "utest/utest_stack_trace.h"
+#include "tls_tests.h"
+
+#if defined(MBEDTLS_SSL_CLI_C) || defined(DOXYGEN_ONLY)
```
Instead of this, can you do
```
#if !defined(MBEDTLS_SSL_CLI_C)
#error [NOT_SUPPORTED] This device does not support SSL library
#endif
```

This `[NOT SUPPORTED]` is magical (undocumented) macro that causes this testbinary to be skipped from build.

michalpasztamobica replied:
Thank you for the hint. I still had to exclude the source code from building, otherwise I was getting more errors and I am afraid CI wouldn't accept that.
I replaced the empty main() with the suggested error macro. Locally it tells me that the build gets skipped:
```
Build successes:
  * ARCH_PRO::GCC_ARM::MBED-BUILD
Build skips:
  * ARCH_PRO::GCC_ARM::TESTS-NETSOCKET-TLS
```
@0xc0170 , will CI accept this as a non-failure or will it flag the build red?

SeppoTakalo replied:
Yes,

Build skips: ... is the correct state.

CI does not try to run those, so it does not cause failures.

michalpasztamobica commented:
Travis was giving some odd errors, so I rebased to latest master to check I am missing some CI fix...

0xc0170 commented:
CI started

0xc0170 commented:
@michalpasztamobica test failures are related

michalpasztamobica commented:
The reason is a known issue in IAR: https://github.com/ARMmbed/mbed-os/issues/8306
I admit, I did not test locally with all compilers. IAR compiler doesn't take the configured heap size into consideration and we have seen in icetea tests that opening two TLSSockets consumes too much memory to fit into the IAR's predefined area size.

@SeppoTakalo , in icetea we left the test failing in order not to hide the bug, but I think for greentea we have to exclude the test, otherwise we will break the public CI. Do you agree?

I wrapped the TLSSOCKET_SIMULTANEOUS test with a preprocessor guard:
```
#ifndef __IAR_SYSTEMS_ICC__
    Case("TLSSOCKET_SIMULTANEOUS", TLSSOCKET_SIMULTANEOUS)
#endif
```
I checked that it doesn't hurt other compilers and IAR simply runs one test less.

michalpasztamobica commented:
@0xc0170 , can we give this another try? Seppo approved my solution and it worked fine locally.

0xc0170 commented:
Ci started

alekla01 commented:
jenkins-ci/greentea-test restarted
